### PR TITLE
Maingate implementation

### PR DESCRIPTION
**Pull request by duguorong009**
Ref: #125

### CODE REVIEW

File: circuit/src/gadgets/main.rs [hunk 1]
```diff
@@ -0,0 +1,950 @@
...
+						+ d.clone() * sd + e * se
+						+ a * b * s_mul_ab + c * d * s_mul_cd
+						+ s_constant),
+			]
+		});
+		selector
+	}
+
+	fn synthesize(
+		self, common: &crate::CommonConfig, selector: &halo2::plonk::Selector,
```
Comment (lazovicff) [inline circuit/src/gadgets/main.rs:117]: Can we import these types on top of the file, before using them here?
Do the same in every other similar case

File: circuit/src/gadgets/main.rs [hunk 2]
```diff
@@ -0,0 +1,950 @@
...
+				let mut ctx = RegionCtx::new(region, 0);
+
+				ctx.enable(*selector)?;
+
+				let result = self
+					.advice
+					.clone()
+					.into_iter()
+					.enumerate()
+					.map(|(i, v)| ctx.assign_advice(common.advice[i], v.value().cloned()))
```
Comment (lazovicff) [inline circuit/src/gadgets/main.rs:132]: This should be copy constrant, not a regular assignment

File: circuit/src/gadgets/main.rs [hunk 3]
```diff
@@ -0,0 +1,950 @@
...
+			},
+		)?;
+
+		// [a, b, c, d, e]
+		let advices = [self.x.clone(), x_inv, res.clone(), zero.clone(), zero];
+		// [s_a, s_b, s_c, s_d, s_e, s_mul_ab, s_mul_cd, s_constant]
+		let fixed =
+			[F::zero(), F::zero(), F::one(), F::zero(), F::zero(), F::one(), F::zero(), -F::one()];
+		let main_chip = MainChip::new(advices, fixed);
+		main_chip.synthesize(common, &config.selector, layouter)?;
```
Comment (lazovicff) [inline circuit/src/gadgets/main.rs:384]: Constraints here are incomplete. Consider the case where:
x = 1
x_inv = 0
res = 1
constrant `x * x_inv + res - 1` would still be satisfied.
We need an additional constraint to ensure that either `x` or `res` are zero

File: circuit/src/gadgets/main.rs [hunk 4]
```diff
@@ -0,0 +1,950 @@
...
+		let diff = layouter.assign_region(
+			|| "assign_values",
+			|region| {
+				let diff = self.x.clone().value().cloned() - self.y.value();
+
+				let mut ctx = RegionCtx::new(region, 0);
+				let diff = ctx.assign_advice(common.advice[0], diff)?;
+				Ok(diff)
+			},
+		)?;
```
Comment (lazovicff) [inline circuit/src/gadgets/main.rs:326]: You are calculating the difference between `x` and `y`, but no constrains are actually applied.
You probably need substraction chipset in here

File: circuit/src/gadgets/main.rs [hunk 5]
```diff
@@ -0,0 +1,950 @@
...
+
+		// [a, b, c, d, e]
+		let advices = [self.bit.clone(), self.x, self.bit.clone(), self.y, res.clone()];
+		// [s_a, s_b, s_c, s_d, s_e, s_mul_ab, s_mul_cd, s_constant]
+		let fixed =
+			[F::zero(), F::zero(), F::zero(), F::one(), -F::one(), F::one(), -F::one(), F::zero()];
+		let main_chip = MainChip::new(advices, fixed);
+		main_chip.synthesize(common, &config.selector, layouter)?;
```
Comment (lazovicff) [inline circuit/src/gadgets/main.rs:444]: We need to constrain `self.bit` to be a boolean

File: circuit/src/gadgets/main.rs (hunk 1 above)
Comment (duguorong009) [reply circuit/src/gadgets/main.rs]: Done in https://github.com/eigen-trust/protocol/pull/132/commits/7fcb9525b5ce9a8ec1d6969b7ca3a616475b1a06

File: circuit/src/gadgets/main.rs (hunk 2 above)
Comment (duguorong009) [reply circuit/src/gadgets/main.rs]: Done in https://github.com/eigen-trust/protocol/pull/132/commits/7fcb9525b5ce9a8ec1d6969b7ca3a616475b1a06

File: circuit/src/gadgets/main.rs (hunk 3 above)
Comment (duguorong009) [reply circuit/src/gadgets/main.rs]: Done in https://github.com/eigen-trust/protocol/pull/132/commits/f152771ee60f10afc1a27411a5e3e7debb0e1d87

File: circuit/src/gadgets/main.rs (hunk 4 above)
Comment (duguorong009) [reply circuit/src/gadgets/main.rs]: Done in https://github.com/eigen-trust/protocol/pull/132/commits/fd67e89d0f674fd64a7f6a4f38e98e67b60db69f

File: circuit/src/gadgets/main.rs (hunk 5 above)
Comment (duguorong009) [reply circuit/src/gadgets/main.rs]: Done in https://github.com/eigen-trust/protocol/pull/132/commits/e9863edb37807f36e4db6ec5440bb31f4760e728

File: circuit/src/gadgets/main.rs [hunk 6]
```diff
@@ -0,0 +1,1004 @@
...
+						} else {
+							*y
+						}
+					},
+				);
+				let mut ctx = RegionCtx::new(region, 0);
+				let res = ctx.assign_advice(common.advice[0], res)?;
+				Ok(res)
+			},
+		)?;
```
Comment (lazovicff) [inline circuit/src/gadgets/main.rs:490]: Can we put `IsBoolChipset` below the region to follow the convention

Comment (duguorong009) [reply circuit/src/gadgets/main.rs]: Done in https://github.com/eigen-trust/protocol/pull/132/commits/fd3fcc62154594ae3dc9eccb0e22db1670f1f798

**Review by lazovicff — APPROVED** [review]

**Review by lazovicff — DISMISSED** [review]

**Review by lazovicff — APPROVED** [review]

**Pull request merged**